hub/core/transform/transform.py

```diff
@@ -0,0 +1,166 @@
+    pipeline_kwargs = pipeline_kwargs[0 : len(pipeline)]
+    pipeline_kwargs += [{}] * (len(pipeline) - len(pipeline_kwargs))
+
+    store(data_in, base_storage, tensors, compute, workers, pipeline, pipeline_kwargs)
+    ds_out.clear_cache()
+    ds_out._load_meta()
+
+
+def store_shard(transform_input: Tuple):
+    """Takes a shard of the original data and iterates through it, producing chunks."""
+    data_shard, size, storage, tensors, pipeline, pipeline_kwargs = transform_input
+
+    # storing the metas in memory to merge later
+    all_index_meta = {key: IndexMeta.create(key, MemoryProvider()) for key in tensors}
+    all_tensor_meta = {key: TensorMeta.create(key, MemoryProvider()) for key in tensors}
+
+    # separate cache for each tensor to prevent frequent flushing, 32 MB ensures only full chunks are written.
+    storage_map = {key: LRUCache(MemoryProvider(), storage, 32 * MB) for key in tensors}
+
+    # will be simply range(len(data_shard)) after AL 1092
+    for i in range(min(len(data_shard), size)):
+        sample = data_shard[i]
+        if isinstance(sample, Dataset):
+            sample_dict = {}
+            for k in sample.tensors:
+                sample_dict[k] = sample[k].numpy()
+            sample = sample_dict
+
+        # always a list of dicts
+        results = transform_sample(sample, pipeline, pipeline_kwargs)
+        for result in results:
+            if set(result.keys()) != tensors:
+                raise TensorMismatchException(list(tensors), list(result.keys()))
+            for key, value in result.items():
+                append_tensor(
+                    value,
+                    key,
+                    storage_map[key],
+                    index_meta=all_index_meta[key],
+                    tensor_meta=all_tensor_meta[key],
+                )
+
+    for tensor in tensors:
+        storage_map[tensor].flush()
+
+    return all_index_meta, all_tensor_meta
+
+
+def store(data_in, storage, tensors, compute, workers, pipeline, pipeline_kwargs):
+    shard_size = math.ceil(len(data_in) / workers)
+    shards = [data_in[i * shard_size : (i + 1) * shard_size] for i in range(workers)]
+
+    # TODO: hacky way to get around improper length of hub dataset slices, can be removed once AL-1092 gets done
+    size_list = [shard_size for _ in range(workers)]
+    extra = shard_size * workers - len(data_in)
+    if size_list:
+        size_list[-1] -= extra
+
+    all_workers_metas = compute.map(
+        store_shard,
+        zip(
+            shards,
+            size_list,
+            repeat(storage),
+            repeat(tensors),
+            repeat(pipeline),
+            repeat(pipeline_kwargs),
+        ),
+    )
+    all_workers_metas = list(all_workers_metas)
+    all_workers_index_meta, all_workers_tensor_meta = zip(*all_workers_metas)
+
+    merge_tensor_metas(all_workers_tensor_meta, storage, tensors)
+    merge_index_metas(all_workers_index_meta, storage, tensors)
```

Comment:
can we move these into `TensorMeta`/`IndexMeta` as non-staticmethods?

Reply:
I thought about it but it didn't really fit in. These are lists of metas that are getting merged, I think a separate util for this is fine.

Reply:
ok no worries